### ISSUE TITLE
[Bug] PPO value function loss is incorrect

### ISSUE THREAD

**Issue by Acciorocketships** [ray-project/ray#19291]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I think the calculation of the clipped value function loss in PPO is incorrect, or at the very least the function of the "vf_clip_param" is misleading.

In the current implementation, `vf_loss1` is the error between the current value function output and the target. `vf_loss_clipped` is the difference between the previous value function and the current value function, clamped with a config clip parameter. `vf_loss2` is the error between `vf_loss_clipped` and the target. Then, the total value loss is computed as the mean of the MAX of `vf_loss1` and `vf_loss2`.

This doesn't make sense, because the clipping parameter really should be _clipping the value of the loss_. However, in the current implementation, the loss can be much bigger than the clipping parameter. For example, even if the clipping parameter is set to 0, then the loss is still either the difference between the current or the last value function output and the target (whichever one is larger). Note that when the clipping parameter is 0 the vf_clipped value simplifies to prev_value_function_out, and when it is inf the vf_clipped value simplifies to value_fn_out.

Let me provide a small example to demonstrate why this implementation doesn't make sense to me. Let's say that the target value is 1.0, and the last value function output was 0.0, but since the last iteration the model improved a bit so the value function output is currently 0.8. Let us also assume that our clipping parameter is 0.1. Then, we would calculate `vf_loss1 = (1.0 - 0.8)**2 = 0.04`,  `vf_loss_clipped = 0.0 + clamp(0.8 - 0.0, 0.1, -0.1) = 0.1`,  `vf_loss2 = (1.0 - 0.1)**2 = 0.81`,  `vf_loss = max(0.04, 0.81) = 0.81`. So, our loss is 0.81 even though we are very close to our target. This will cause us to overshoot. As you can see, the clip parameter did nothing to actually clamp the value function loss, and the max operation caused us to choose a loss that was far too large.

In contrast, I would expect the vf loss to be calculated by simply clamping `vf_loss1`:
`vf_loss = clamp( (value_target - curr_value_function_out)**2, clip_param, -clip_param)`

cc @smorad 

### Reproduction script

from `ppo_torch_policy.py:88`:
```python
prev_value_fn_out = train_batch[SampleBatch.VF_PREDS]
value_fn_out = model.value_function()
vf_loss1 = torch.pow(
            value_fn_out - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_clipped = prev_value_fn_out + torch.clamp(
            value_fn_out - prev_value_fn_out, -policy.config["vf_clip_param"],
            policy.config["vf_clip_param"])
vf_loss2 = torch.pow(
vf_clipped - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_loss = torch.max(vf_loss1, vf_loss2)
mean_vf_loss = reduce_mean_valid(vf_loss)
```

### Anything else

proposed updated code:
```python
prev_value_fn_out = train_batch[SampleBatch.VF_PREDS]
value_fn_out = model.value_function()
vf_loss1 = torch.pow(
            value_fn_out - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_loss = torch.clamp(
            vf_loss1, -policy.config["vf_clip_param"],
            policy.config["vf_clip_param"])
mean_vf_loss = reduce_mean_valid(vf_loss)
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by vwxyzjn** [comment]
@mvindiola1 thanks for reporting. For reference, the original implementation looks like this:

```python
# Clip the value to reduce variability during Critic training
# Get the predicted value
vpred = train_model.vf
vpredclipped = OLDVPRED + tf.clip_by_value(train_model.vf - OLDVPRED, - CLIPRANGE, CLIPRANGE)
# Unclipped value
vf_losses1 = tf.square(vpred - R)
# Clipped value
vf_losses2 = tf.square(vpredclipped - R)
vf_loss = .5 * tf.reduce_mean(tf.maximum(vf_losses1, vf_losses2))
```

https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/model.py#L68-L75

**Comment by Acciorocketships** [comment]
There is currently a way to do that---if you set the vf_clipping_param to inf then the loss simplifies to vf_loss1, which is the unclipped version. However, I still think we should look into this.
